### ISSUE TITLE
Use static copyright years

### ISSUE THREAD

**Issue by hugovk** [python/cpython#126133]
# Short version

Each January we update the PSF copyright years in the repo:

```
Copyright © 2001-2024 Python Software Foundation. All rights reserved.
```

2025 is [approaching](https://hugovk.github.io/year-progress-bar/).

Can we get permission from the PSF to adopt a copyright that does not need updating every year?

I suggest something like one of the following:

```
Copyright © Python Software Foundation. All rights reserved.
Copyright © 2001 Python Software Foundation. All rights reserved.
Copyright © 2001-present Python Software Foundation. All rights reserved.
```

# Longer version

Last year, we updated it in 15 places in 10 files:

* 2024 https://github.com/python/cpython/pull/113608

Sometimes we miss a few and have to update them later:

* 2023 https://github.com/python/cpython/pull/100848
       https://github.com/python/cpython/pull/100859

* 2022 https://github.com/python/cpython/pull/30335

Or we miss something, merge the fix a year late, which needs refixing:

* 2021 https://github.com/python/cpython/pull/24036
       https://github.com/python/cpython/pull/25514
       https://github.com/python/cpython/pull/31347
* 2020 https://github.com/python/cpython/pull/17801
* 2019 https://github.com/python/cpython/pull/11404
* 2018 https://github.com/python/cpython/pull/5094
* 2017 https://github.com/python/cpython/pull/3421

And sometimes we backport these to (some) older branches, but other years don't:

* 2023 https://github.com/python/cpython/pull/100849
       https://github.com/python/cpython/pull/100850
       https://github.com/python/cpython/pull/100851
       https://github.com/python/cpython/pull/100852
       https://github.com/python/cpython/pull/100853
* 2022 https://github.com/python/cpython/pull/30336
       https://github.com/python/cpython/pull/30337
* 2020 https://github.com/python/cpython/pull/17802
       https://github.com/python/cpython/pull/17803
       https://github.com/python/cpython/pull/17804
       https://github.com/python/cpython/pull/17805
* 2021 https://github.com/python/cpython/pull/24045
       https://github.com/python/cpython/pull/24046
       https://github.com/python/cpython/pull/24047
       https://github.com/python/cpython/pull/24052
       https://github.com/python/cpython/pull/24054
* 2019 https://github.com/python/cpython/pull/11406
       https://github.com/python/cpython/pull/11407
       https://github.com/python/cpython/pull/11408
* 2018 https://github.com/python/cpython/pull/5104
       https://github.com/python/cpython/pull/5105

Sometimes there's lots of duplicate PRs that end up closed and wasting everyone's time:

* https://github.com/python/cpython/pull/6723
* https://github.com/python/cpython/pull/5069
* https://github.com/python/cpython/pull/76
* https://github.com/python/cpython/pull/62
* https://github.com/python/cpython/pull/56
* https://github.com/python/cpython/pull/4

Not only is this tedious work, it is likely unnecessary. A lot of big projects have stopped updating copyright years: https://hynek.me/til/copyright-years/ 

For example:

```
Copyright (c) 2009 The Go Authors. All rights reserved.
Copyright (c) 2015 - present Microsoft Corporation
Copyright 2013 Netflix, Inc.
Copyright (c) Meta Platforms, Inc. and affiliates.
Copyright (c) Microsoft Corporation. All rights reserved.
Copyright (c) .NET Foundation and Contributors
```

These either have only have the first year, omit years altogether, or end with "present".

Can we do something similar?

### Linked PRs
* gh-126236

**Comment by Wulian233** [comment]
+1 I also suggestion deleting the years of PSF 2001-20xx in other files. This should always be up to date. Currently, files with these rows are searched (200x 201x) over nearly 10~20 years. I don't think they need to repeat the statement because it's already stipulated in the LICENSE. But that's a lot of files to change

2001-200x
https://github.com/search?q=repo:python/cpython 2001-200&type=code

2001-201x
https://github.com/search?q=repo:python/cpython 2001-201&type=code

**Comment by gpshead** [comment]
IMNSHO lets just adopt this already.  Do not update existing copyright dates in existing files.

Ex: From other your examples above - Google's policy was always "never touch the year" with the year of file creation or copyright statement addition being the standard. As noted in https://opensource.google/documentation/reference/releasing/preparing#license-headers

**Comment by AA-Turner** [comment]
From that draft PR -- seeing the version with '2024' removed as the end year leads me to slightly prefer the 'YYYY - present' option, especially in lists of several prior copyright assignments. Hugo's option 1 (no year) would also render this moot, though is slightly less consistent.

I would also suggest removing copyright comments from modules/files where the PSF is the only cited copyright holder, as this is simply duplicative information.

A

**Comment by gpshead** [comment]
> I would also suggest removing copyright comments from modules/files where the PSF is the only cited copyright holder, as this is simply duplicative information.

I wouldn't _remove_ entire copyright statements from anywhere.

Hugo should reach out to the PSF (as copyright holder) for advice.  Our goal is to simplify our life, have a policy to point to, and reduce toil.

We should codify said advice in our dev guide and simply reply to anyone attempting to modify copyright statements with a link to that documentation.

**Comment by hugovk** [comment]
I've emailed PSF legal.

**Comment by hugovk** [comment]
And the reply:

---

The first thing to understand is that copyright notices are optional/informational. There was a time that copyright owners were required to include a notice in order for their copyrights to be protected; by international treaty, this requirement has been abolished. So the form of your copyright notices will not impact the rights of the copyright owner.

Copyright notices can, however, still perform a useful informative role. According to the US copyright office, the elements of a valid copyright notice are:

- The copyright symbol ©; the word “copyright”; or the abbreviation “copr.”;
- The year of first publication of the work; and
- The name of the copyright owner.

Example: © 2024 John Doe

For a work that's published only once, the year of first publication is a simple matter. A file in an open source project, on the other hand, is essentially an accretion of revisions or derivative works, each with different years of first publication. This is why you see multiple dates or date ranges in the corresponding copyright notices. For example, if a file was created in 2015 and revised in 2017, 2018, 2019, and 2022, you might render the copyright notice in one of the following ways:

- © 2015, 2017-2019, 2022 John Doe
- © 2015, 2017, 2018, 2019, 2022 John Doe

It's probably fairly common in such cases to use "2015-2022" as the date range, although this provides somewhat less information about the dates of first publication of the various revisions, and arguably implies incorrectly that the file was revised in each of the years in the range.

It doesn't make particular sense to update notices to add the present year at the end of the date range in *every* copyright notice, regardless of whether the corresponding file has been updated recently. But it's also not going to impact anyone's rights, nor is using "-present" instead.

So I think both of the formats you propose that include the year would be fine. I would not recommend leaving out the year of first publication entirely, as it's one of the elements of a valid copyright notice.

I'll also note that "Python Software Foundation" is probably not an accurate statement of the author(s) of PSF projects, since PSF does not (as far as I know) require copyright assignment. That said, it's also impractical and error-prone to include notices for all of the authors of a file. Some common alternatives are "FOO Project Contributors" or "The authors listed in CONTRIBUTORS." None of these are ideal, because they either don't actually name the authors, or they may become inaccurate as an author's contributions disappear over time.

In short, there's no ideal solution, but thankfully notices aren't that important. Pick something reasonable and do your best to stick to it.

**Comment by hugovk** [comment]
Therefore I propose https://github.com/python/cpython/pull/126236 as it is. (Left as draft to avoid pinging the CODEOWNERS.)

> We should codify said advice in our dev guide and simply reply to anyone attempting to modify copyright statements with a link to that documentation.

Good idea. Let's do this after merge.

**Comment by hugovk** [comment]
I've merged https://github.com/python/cpython/pull/126236 and opened https://github.com/python/devguide/pull/1470.

**Comment by malemburg** [comment]
@hugovk: Before making such changes and releasing the distributions with these changes, please ask the PSF Board for approval. They are the only ones who have the authority to change license texts and possibly even PSF copyright notices (since they are being referred to in the PSF license). Thanks.

**Comment by eximious** [comment]
The PSF Legal list is monitored by PSF staff (including myself) and is the correct path for these kinds of issues. Topics that need full board involvement get sent up and ones that don't need full board involvement get decided there, either without or without legal counsel depending on the topic. In this particular instance, we did loop in counsel to make sure that the PSF is fulfilling our legal duty as copyright holders.

**Comment by webknjaz** [comment]
Apparently, python.org still uses the current year: https://github.com/python/pythondotorg/pull/597/files#diff-008dcb3426febd767787b1521f1fe33086313b927ea37eaab86df5fa88a51698R53.
So does pypi.org: https://github.com/pypi/warehouse/issues/524 / https://github.com/pypi/warehouse/pull/898/files#diff-8fc59fb79d6d1c00f702c99239e388a627fc51137f813d9e2be1d078d0a50147R111 / https://github.com/pypi/warehouse/pull/898/files#diff-8fc59fb79d6d1c00f702c99239e388a627fc51137f813d9e2be1d078d0a50147R111.

Should those be updated too?

**Comment by hugovk** [comment]
I don't have merge rights for those, but you can open issues or PRs over there to ask their maintainers.
